// lerna/lib/commands/ExecCommand.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = exports.builder = exports.describe = exports.command = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

exports.handler = handler;

var _ChildProcessUtilities = require("../ChildProcessUtilities");

var _ChildProcessUtilities2 = _interopRequireDefault(_ChildProcessUtilities);

var _PackageUtilities = require("../PackageUtilities");

var _PackageUtilities2 = _interopRequireDefault(_PackageUtilities);

var _Command2 = require("../Command");

var _Command3 = _interopRequireDefault(_Command2);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _possibleConstructorReturn(self, call) { if (!self) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return call && (typeof call === "object" || typeof call === "function") ? call : self; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function, not " + typeof superClass); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, enumerable: false, writable: true, configurable: true } }); if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass; }

function _toConsumableArray(arr) { if (Array.isArray(arr)) { for (var i = 0, arr2 = Array(arr.length); i < arr.length; i++) { arr2[i] = arr[i]; } return arr2; } else { return Array.from(arr); } }

function handler(argv) {
  return new ExecCommand([argv.command].concat(_toConsumableArray(argv.args)), argv).run();
}

var command = exports.command = "exec <command> [args..]";

var describe = exports.describe = "Run an arbitrary command in each package.";

var builder = exports.builder = {};

var ExecCommand = function (_Command) {
  _inherits(ExecCommand, _Command);

  function ExecCommand() {
    _classCallCheck(this, ExecCommand);

    return _possibleConstructorReturn(this, (ExecCommand.__proto__ || Object.getPrototypeOf(ExecCommand)).apply(this, arguments));
  }

  _createClass(ExecCommand, [{
    key: "initialize",
    value: function initialize(callback) {
      this.command = this.input[0];
      this.args = this.input.slice(1);

      if (!this.command) {
        callback(new Error("You must specify which command to run."));
        return;
      }

      this.batchedPackages = this.toposort ? _PackageUtilities2.default.topologicallyBatchPackages(this.filteredPackages, { logger: this.logger }) : [this.filteredPackages];

      callback(null, true);
    }
  }, {
    key: "execute",
    value: function execute(callback) {
      var _this2 = this;

      _PackageUtilities2.default.runParallelBatches(this.batchedPackages, function (pkg) {
        return function (done) {
          _this2.runCommandInPackage(pkg, done);
        };
      }, this.concurrency, callback);
    }
  }, {
    key: "getOpts",
    value: function getOpts(pkg) {
      return {
        cwd: pkg.location,
        shell: true,
        env: Object.assign({}, process.env, {
          LERNA_PACKAGE_NAME: pkg.name
        })
      };
    }
  }, {
    key: "runCommandInPackage",
    value: function runCommandInPackage(pkg, callback) {
      var _this3 = this;

      _ChildProcessUtilities2.default.spawn(this.command, this.args, this.getOpts(pkg), function (err) {
        if (err && err.code) {
          _this3.logger.error(`Errored while executing '${err.cmd}' in '${pkg.name}'`);
        }
        callback(err);
      });
    }
  }]);

  return ExecCommand;
}(_Command3.default);

exports.default = ExecCommand;
