// lerna/lib/FileSystemUtilities.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = undefined;

var _createClass = function () { function defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } } return function (Constructor, protoProps, staticProps) { if (protoProps) defineProperties(Constructor.prototype, protoProps); if (staticProps) defineProperties(Constructor, staticProps); return Constructor; }; }();

var _dec, _dec2, _dec3, _dec4, _dec5, _dec6, _dec7, _dec8, _dec9, _dec10, _dec11, _dec12, _dec13, _desc, _value, _class;

var _fsPromise = require("fs-promise");

var _fsPromise2 = _interopRequireDefault(_fsPromise);

var _pathExists = require("path-exists");

var _pathExists2 = _interopRequireDefault(_pathExists);

var _logger = require("./logger");

var _logger2 = _interopRequireDefault(_logger);

var _cmdShim = require("cmd-shim");

var _cmdShim2 = _interopRequireDefault(_cmdShim);

var _readCmdShim = require("read-cmd-shim");

var _readCmdShim2 = _interopRequireDefault(_readCmdShim);

var _path = require("path");

var _path2 = _interopRequireDefault(_path);

var _ChildProcessUtilities = require("./ChildProcessUtilities");

var _ChildProcessUtilities2 = _interopRequireDefault(_ChildProcessUtilities);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _applyDecoratedDescriptor(target, property, decorators, descriptor, context) {
  var desc = {};
  Object['ke' + 'ys'](descriptor).forEach(function (key) {
    desc[key] = descriptor[key];
  });
  desc.enumerable = !!desc.enumerable;
  desc.configurable = !!desc.configurable;

  if ('value' in desc || desc.initializer) {
    desc.writable = true;
  }

  desc = decorators.slice().reverse().reduce(function (desc, decorator) {
    return decorator(target, property, desc) || desc;
  }, desc);

  if (context && desc.initializer !== void 0) {
    desc.value = desc.initializer ? desc.initializer.call(context) : void 0;
    desc.initializer = undefined;
  }

  if (desc.initializer === void 0) {
    Object['define' + 'Property'](target, property, desc);
    desc = null;
  }

  return desc;
}

var ENDS_WITH_NEW_LINE = /\n$/;

function ensureEndsWithNewLine(string) {
  return ENDS_WITH_NEW_LINE.test(string) ? string : string + "\n";
}

// NOTE: if rimraf moves the location of its executable, this will need to be updated
var RIMRAF_CLI = require.resolve("rimraf/bin");

// globs only return directories with a trailing slash
function trailingSlash(filePath) {
  return _path2.default.normalize(`${filePath}/`);
}

var FileSystemUtilities = (_dec = _logger2.default.logifyAsync(), _dec2 = _logger2.default.logifySync(), _dec3 = _logger2.default.logifySync(), _dec4 = _logger2.default.logifyAsync(), _dec5 = _logger2.default.logifyAsync(), _dec6 = _logger2.default.logifySync(), _dec7 = _logger2.default.logifySync(), _dec8 = _logger2.default.logifySync(), _dec9 = _logger2.default.logifySync(), _dec10 = _logger2.default.logifyAsync(), _dec11 = _logger2.default.logifyAsync(), _dec12 = _logger2.default.logifySync(), _dec13 = _logger2.default.logifySync(), (_class = function () {
  function FileSystemUtilities() {
    _classCallCheck(this, FileSystemUtilities);
  }

  _createClass(FileSystemUtilities, null, [{
    key: "mkdirp",
    value: function mkdirp(filePath, callback) {
      _fsPromise2.default.ensureDir(filePath, callback);
    }
  }, {
    key: "readdirSync",
    value: function readdirSync(filePath) {
      return _fsPromise2.default.readdirSync(filePath);
    }
  }, {
    key: "existsSync",
    value: function existsSync(filePath) {
      return _pathExists2.default.sync(filePath);
    }
  }, {
    key: "writeFile",
    value: function writeFile(filePath, fileContents, callback) {
      _fsPromise2.default.writeFile(filePath, ensureEndsWithNewLine(fileContents), callback);
    }
  }, {
    key: "rename",
    value: function rename(from, to, callback) {
      _fsPromise2.default.rename(from, to, callback);
    }
  }, {
    key: "renameSync",
    value: function renameSync(from, to) {
      _fsPromise2.default.renameSync(from, to);
    }
  }, {
    key: "writeFileSync",
    value: function writeFileSync(filePath, fileContents) {
      _fsPromise2.default.writeFileSync(filePath, ensureEndsWithNewLine(fileContents));
    }
  }, {
    key: "readFileSync",
    value: function readFileSync(filePath) {
      return _fsPromise2.default.readFileSync(filePath, "utf8").trim();
    }
  }, {
    key: "statSync",
    value: function statSync(filePath) {
      return _fsPromise2.default.statSync(filePath);
    }
  }, {
    key: "rimraf",
    value: function rimraf(dirPaths, callback) {
      // Shelling out to a child process for a noop is expensive.
      // Checking if `dirPath` exists to be removed is cheap.
      // This lets us short-circuit if we don't have anything to do.
      var mapper = function mapper(dir) {
        return Promise.all([(0, _pathExists2.default)(dir), dir]);
      };

      return Promise.all(dirPaths.map(mapper)).then(function (values) {
        return values.filter(function (x) {
          return x[0];
        }).map(function (x) {
          return x[1];
        });
      }).then(function (candidates) {
        if (!candidates.length) {
          return callback();
        }

        var args = candidates.map(trailingSlash);
        args.unshift("--no-glob");
        args.unshift(RIMRAF_CLI);

        // We call this resolved CLI path in the "path/to/node path/to/cli <..args>"
        // pattern to avoid Windows hangups with shebangs (e.g., WSH can't handle it)
        return _ChildProcessUtilities2.default.spawn(process.execPath, args, {}, callback);
      });
    }
  }, {
    key: "symlink",
    value: function symlink(src, dest, type, callback) {
      if (process.platform === "win32") {
        createWindowsSymlink(src, dest, type, callback);
      } else {
        createPosixSymlink(src, dest, type, callback);
      }
    }
  }, {
    key: "unlinkSync",
    value: function unlinkSync(filePath) {
      _fsPromise2.default.unlinkSync(filePath);
    }
  }, {
    key: "isSymlink",
    value: function isSymlink(filePath) {
      var result = void 0;

      if (process.platform === "win32") {
        result = resolveWindowsSymlink(filePath);
      } else {
        result = resolvePosixSymlink(filePath);
      }

      return result;
    }
  }]);

  return FileSystemUtilities;
}(), (_applyDecoratedDescriptor(_class, "mkdirp", [_dec], Object.getOwnPropertyDescriptor(_class, "mkdirp"), _class), _applyDecoratedDescriptor(_class, "readdirSync", [_dec2], Object.getOwnPropertyDescriptor(_class, "readdirSync"), _class), _applyDecoratedDescriptor(_class, "existsSync", [_dec3], Object.getOwnPropertyDescriptor(_class, "existsSync"), _class), _applyDecoratedDescriptor(_class, "writeFile", [_dec4], Object.getOwnPropertyDescriptor(_class, "writeFile"), _class), _applyDecoratedDescriptor(_class, "rename", [_dec5], Object.getOwnPropertyDescriptor(_class, "rename"), _class), _applyDecoratedDescriptor(_class, "renameSync", [_dec6], Object.getOwnPropertyDescriptor(_class, "renameSync"), _class), _applyDecoratedDescriptor(_class, "writeFileSync", [_dec7], Object.getOwnPropertyDescriptor(_class, "writeFileSync"), _class), _applyDecoratedDescriptor(_class, "readFileSync", [_dec8], Object.getOwnPropertyDescriptor(_class, "readFileSync"), _class), _applyDecoratedDescriptor(_class, "statSync", [_dec9], Object.getOwnPropertyDescriptor(_class, "statSync"), _class), _applyDecoratedDescriptor(_class, "rimraf", [_dec10], Object.getOwnPropertyDescriptor(_class, "rimraf"), _class), _applyDecoratedDescriptor(_class, "symlink", [_dec11], Object.getOwnPropertyDescriptor(_class, "symlink"), _class), _applyDecoratedDescriptor(_class, "unlinkSync", [_dec12], Object.getOwnPropertyDescriptor(_class, "unlinkSync"), _class), _applyDecoratedDescriptor(_class, "isSymlink", [_dec13], Object.getOwnPropertyDescriptor(_class, "isSymlink"), _class)), _class));
exports.default = FileSystemUtilities;


function createSymbolicLink(src, dest, type, callback) {
  _fsPromise2.default.lstat(dest, function (err) {
    if (!err) {
      // Something exists at `dest`.  Need to remove it first.
      _fsPromise2.default.unlink(dest, function () {
        return _fsPromise2.default.symlink(src, dest, type, callback);
      });
    } else {
      _fsPromise2.default.symlink(src, dest, type, callback);
    }
  });
}

function createPosixSymlink(origin, dest, type, callback) {
  if (type === "exec") {
    type = "file";
  }
  var src = _path2.default.relative(_path2.default.dirname(dest), origin);
  createSymbolicLink(src, dest, type, callback);
}

function createWindowsSymlink(src, dest, type, callback) {
  if (type === "exec") {
    (0, _cmdShim2.default)(src, dest, callback);
  } else {
    createSymbolicLink(src, dest, type, callback);
  }
}

function resolveSymbolicLink(filePath) {
  var lstat = _fsPromise2.default.lstatSync(filePath);
  var isSymlink = lstat.isSymbolicLink() ? _path2.default.resolve(_path2.default.dirname(filePath), _fsPromise2.default.readlinkSync(filePath)) : false;

  return {
    isSymlink,
    lstat
  };
}

function resolvePosixSymlink(filePath) {
  var _resolveSymbolicLink = resolveSymbolicLink(filePath),
      isSymlink = _resolveSymbolicLink.isSymlink;

  return isSymlink;
}

function resolveWindowsSymlink(filePath) {
  var _resolveSymbolicLink2 = resolveSymbolicLink(filePath),
      isSymlink = _resolveSymbolicLink2.isSymlink,
      lstat = _resolveSymbolicLink2.lstat;

  if (lstat.isFile() && !isSymlink) {
    try {
      return _path2.default.resolve(_path2.default.dirname(filePath), _readCmdShim2.default.sync(filePath));
    } catch (e) {
      return false;
    }
  }

  return isSymlink && _path2.default.resolve(isSymlink);
}
module.exports = exports["default"];
